### PR TITLE
Add distinct parameter to API endpoints

### PR DESCRIPTION
**Pull request by GGP1**
|Related issue|
|---|
|Closes https://github.com/wazuh/wazuh/issues/16715|

## Description

Added the `distinct` parameter to several endpoints, updated their tests and specifications.

## Tests

<details>
  <summary>test_agent_GET_endpoints</summary>

```console
(venv) gasti@pop-os:~/work/wazuh/api/test/integration$ pytest -vv test_agent_GET_endpoints.tavern.yaml
=================================================== test session starts ====================================================
platform linux -- Python 3.9.16, pytest-7.3.1, pluggy-1.2.0 -- /home/gasti/work/wazuh/venv/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.9.16', 'Platform': 'Linux-6.2.0-76060200-generic-x86_64-with-glibc2.35', 'Packages': {'pytest': '7.3.1', 'pluggy': '1.2.0'}, 'Plugins': {'asyncio': '0.18.1', 'tavern': '1.23.5', 'trio': '0.7.0', 'html': '2.1.1', 'aiohttp': '1.0.4', 'metadata': '3.0.0'}}
rootdir: /home/gasti/work/wazuh/api/test/integration
configfile: pytest.ini
plugins: asyncio-0.18.1, tavern-1.23.5, trio-0.7.0, html-2.1.1, aiohttp-1.0.4, metadata-3.0.0
asyncio: mode=auto
collected 93 items                                                                                                         

test_agent_GET_endpoints.tavern.yaml::GET /agents PASSED                                                             [  1%]
test_agent_GET_endpoints.tavern.yaml::GET /agents with single agent id PASSED                                        [  2%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/{agent_id}/config/{component}/{configuration} PASSED               [  3%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/{agent_id}/config/{component}/{configuration} (only cluster configuration) PASSED [  4%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/{agent_id}/group/is_sync PASSED                                    [  5%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/{agent_id}/key PASSED                                              [  6%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/{agent_id}/daemons/stats PASSED                                    [  7%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/{agent_id}/stats/{component} PASSED                                [  8%]
test_agent_GET_endpoints.tavern.yaml::GET /groups PASSED                                                             [  9%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[md5] PASSED                                            [ 10%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[sha1] PASSED                                           [ 11%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[sha224] PASSED                                         [ 12%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[sha256] PASSED                                         [ 13%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[sha384] PASSED                                         [ 15%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[sha512] PASSED                                         [ 16%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[blake2b] PASSED                                        [ 17%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[blake2s] PASSED                                        [ 18%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[sha3_224] PASSED                                       [ 19%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[sha3_256] PASSED                                       [ 20%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[sha3_384] PASSED                                       [ 21%]
test_agent_GET_endpoints.tavern.yaml::GET /groups filter hash[sha3_512] PASSED                                       [ 22%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents PASSED                                           [ 23%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[configSum] PASSED                  [ 24%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[dateAdd] PASSED                    [ 25%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[group] PASSED                      [ 26%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[id] PASSED                         [ 27%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[ip] PASSED                         [ 29%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[lastKeepAlive] PASSED              [ 30%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[manager] PASSED                    [ 31%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[mergedSum] PASSED                  [ 32%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[name] PASSED                       [ 33%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[node_name] PASSED                  [ 34%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[os.arch] PASSED                    [ 35%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[os.build] PASSED                   [ 36%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[os.codename] PASSED                [ 37%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[os.major] PASSED                   [ 38%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[os.minor] PASSED                   [ 39%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[os.name] PASSED                    [ 40%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[os.platform] PASSED                [ 41%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[os.uname] PASSED                   [ 43%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[os.version] PASSED                 [ 44%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[registerIP] PASSED                 [ 45%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[status] PASSED                     [ 46%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/agents {sort,select}[version] PASSED                    [ 47%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/configuration PASSED                                    [ 48%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files PASSED                                            [ 49%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[md5] PASSED                           [ 50%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[sha1] PASSED                          [ 51%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[sha224] PASSED                        [ 52%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[sha256] PASSED                        [ 53%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[sha384] PASSED                        [ 54%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[sha512] PASSED                        [ 55%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[blake2b] PASSED                       [ 56%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[blake2s] PASSED                       [ 58%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[sha3_224] PASSED                      [ 59%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[sha3_256] PASSED                      [ 60%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[sha3_384] PASSED                      [ 61%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files filter hash[sha3_512] PASSED                      [ 62%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files/{filename}/json PASSED                            [ 63%]
test_agent_GET_endpoints.tavern.yaml::GET /groups/{group_id}/files/{filename}/xml PASSED                             [ 64%]
test_agent_GET_endpoints.tavern.yaml::GET /agents?name=agent_name PASSED                                             [ 65%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/no_group PASSED                                                    [ 66%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/no_group {sort,select}[dateAdd] PASSED                             [ 67%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/no_group {sort,select}[id] PASSED                                  [ 68%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/no_group {sort,select}[ip] PASSED                                  [ 69%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/no_group {sort,select}[name] PASSED                                [ 70%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/no_group {sort,select}[node_name] PASSED                           [ 72%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/no_group {sort,select}[registerIP] PASSED                          [ 73%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/no_group {sort,select}[status] PASSED                              [ 74%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/outdated PASSED                                                    [ 75%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct PASSED                                              [ 76%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[dateAdd] PASSED                            [ 77%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[id] PASSED                                 [ 78%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[ip] PASSED                                 [ 79%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[lastKeepAlive] PASSED                      [ 80%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[manager] PASSED                            [ 81%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[name] PASSED                               [ 82%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[node_name] PASSED                          [ 83%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[os.arch] PASSED                            [ 84%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[os.build] PASSED                           [ 86%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[os.codename] PASSED                        [ 87%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[os.major] PASSED                           [ 88%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[os.minor] PASSED                           [ 89%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[os.name] PASSED                            [ 90%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[os.platform] PASSED                        [ 91%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[os.uname] PASSED                           [ 92%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[os.version] PASSED                         [ 93%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[registerIP] PASSED                         [ 94%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[status] PASSED                             [ 95%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/stats/distinct {fields}[version] PASSED                            [ 96%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/summary/status PASSED                                              [ 97%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/summary/os PASSED                                                  [ 98%]
test_agent_GET_endpoints.tavern.yaml::GET /agents/upgrade_result PASSED                                              [100%]

======================================== 93 passed, 2 warnings in 207.83s (0:03:27) ========================================
```
</details>

<details>
  <summary>test_decoder_endpoints</summary>

```console
(venv) gasti@pop-os:~/work/wazuh/api/test/integration$ pytest -vv test_decoder_endpoints.tavern.yaml
==================================================== test session starts ====================================================
platform linux -- Python 3.9.16, pytest-7.3.1, pluggy-1.2.0 -- /home/gasti/work/wazuh/venv/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.9.16', 'Platform': 'Linux-6.2.0-76060200-generic-x86_64-with-glibc2.35', 'Packages': {'pytest': '7.3.1', 'pluggy': '1.2.0'}, 'Plugins': {'asyncio': '0.18.1', 'tavern': '1.23.5', 'trio': '0.7.0', 'html': '2.1.1', 'aiohttp': '1.0.4', 'metadata': '3.0.0'}}
rootdir: /home/gasti/work/wazuh/api/test/integration
configfile: pytest.ini
plugins: asyncio-0.18.1, tavern-1.23.5, trio-0.7.0, html-2.1.1, aiohttp-1.0.4, metadata-3.0.0
asyncio: mode=auto
collected 25 items                                                                                                          

test_decoder_endpoints.tavern.yaml::GET /decoders PASSED                                                              [  4%]
test_decoder_endpoints.tavern.yaml::GET /decoders[filename] PASSED                                                    [  8%]
test_decoder_endpoints.tavern.yaml::GET /decoders[relative_dirname] PASSED                                            [ 12%]
test_decoder_endpoints.tavern.yaml::GET /decoders[name] PASSED                                                        [ 16%]
test_decoder_endpoints.tavern.yaml::GET /decoders[position] PASSED                                                    [ 20%]
test_decoder_endpoints.tavern.yaml::GET /decoders[status] PASSED                                                      [ 24%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name} PASSED                                               [ 28%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[filename] PASSED                                     [ 32%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[relative_dirname] PASSED                             [ 36%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[name] PASSED                                         [ 40%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[position] PASSED                                     [ 44%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[status] PASSED                                       [ 48%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files PASSED                                                        [ 52%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files[filename] PASSED                                              [ 56%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files[relative_dirname] PASSED                                      [ 60%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files[status] PASSED                                                [ 64%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents PASSED                                                      [ 68%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[filename] PASSED                                            [ 72%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[name] PASSED                                                [ 76%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[relative_dirname] PASSED                                    [ 80%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[position] PASSED                                            [ 84%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[status] PASSED                                              [ 88%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files/{file} PASSED                                                 [ 92%]
test_decoder_endpoints.tavern.yaml::PUT /rule/decoders/{filename} PASSED                                              [ 96%]
test_decoder_endpoints.tavern.yaml::DELETE /decoders/files/{filename} PASSED                                          [100%]
======================================== 25 passed, 2 warnings in 417.16s (0:06:57) =========================================
```
</details>

<details>
  <summary>test_mitre_endpoints</summary>

```console
(venv) gasti@pop-os:~/work/wazuh/api/test/integration$ pytest -vv test_mitre_endpoints.tavern.yaml
=================================================== test session starts ====================================================
platform linux -- Python 3.9.16, pytest-7.3.1, pluggy-1.2.0 -- /home/gasti/work/wazuh/venv/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.9.16', 'Platform': 'Linux-6.2.0-76060200-generic-x86_64-with-glibc2.35', 'Packages': {'pytest': '7.3.1', 'pluggy': '1.2.0'}, 'Plugins': {'asyncio': '0.18.1', 'tavern': '1.23.5', 'trio': '0.7.0', 'html': '2.1.1', 'aiohttp': '1.0.4', 'metadata': '3.0.0'}}
rootdir: /home/gasti/work/wazuh/api/test/integration
configfile: pytest.ini
plugins: asyncio-0.18.1, tavern-1.23.5, trio-0.7.0, html-2.1.1, aiohttp-1.0.4, metadata-3.0.0
asyncio: mode=auto
collected 7 items                                                                                                          

test_mitre_endpoints.tavern.yaml::GET /mitre/metadata PASSED                                                         [ 14%]
test_mitre_endpoints.tavern.yaml::GET /mitre/mitigations PASSED                                                      [ 28%]
test_mitre_endpoints.tavern.yaml::GET /mitre/references PASSED                                                       [ 42%]
test_mitre_endpoints.tavern.yaml::GET /mitre/tactics PASSED                                                          [ 57%]
test_mitre_endpoints.tavern.yaml::GET /mitre/techniques PASSED                                                       [ 71%]
test_mitre_endpoints.tavern.yaml::GET /mitre/groups PASSED                                                           [ 85%]
test_mitre_endpoints.tavern.yaml::GET /mitre/software PASSED                                                         [100%]

======================================== 7 passed, 2 warnings in 124.72s (0:02:04) =========================================
```
</details>

<details>
  <summary>test_manager_endpoints</summary>

```console
(venv) gasti@pop-os:~/work/wazuh/api/test/integration$ pytest -vv test_manager_endpoints.tavern.yaml
=================================================== test session starts ====================================================
platform linux -- Python 3.9.16, pytest-7.3.1, pluggy-1.2.0 -- /home/gasti/work/wazuh/venv/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.9.16', 'Platform': 'Linux-6.2.0-76060200-generic-x86_64-with-glibc2.35', 'Packages': {'pytest': '7.3.1', 'pluggy': '1.2.0'}, 'Plugins': {'asyncio': '0.18.1', 'tavern': '1.23.5', 'trio': '0.7.0', 'html': '2.1.1', 'aiohttp': '1.0.4', 'metadata': '3.0.0'}}
rootdir: /home/gasti/work/wazuh/api/test/integration
configfile: pytest.ini
plugins: asyncio-0.18.1, tavern-1.23.5, trio-0.7.0, html-2.1.1, aiohttp-1.0.4, metadata-3.0.0
asyncio: mode=auto
collected 32 items                                                                                                         

test_manager_endpoints.tavern.yaml::GET /manager/status PASSED                                                       [  3%]
test_manager_endpoints.tavern.yaml::GET /manager/info PASSED                                                         [  6%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[alerts] PASSED                              [  9%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[auth] PASSED                                [ 12%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[cis-cat] PASSED                             [ 15%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[cluster] PASSED                             [ 18%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[command] PASSED                             [ 21%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[global] PASSED                              [ 25%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[localfile] PASSED                           [ 28%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[osquery] PASSED                             [ 31%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[remote] PASSED                              [ 34%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[rootcheck] PASSED                           [ 37%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[ruleset] PASSED                             [ 40%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[sca] PASSED                                 [ 43%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[syscheck] PASSED                            [ 46%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[syscollector] PASSED                        [ 50%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration?{section}[vulnerability-detector] PASSED              [ 53%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration PASSED                                                [ 56%]
test_manager_endpoints.tavern.yaml::GET /manager/daemons/stats PASSED                                                [ 59%]
test_manager_endpoints.tavern.yaml::GET /manager/stats PASSED                                                        [ 62%]
test_manager_endpoints.tavern.yaml::GET /manager/stats/hourly PASSED                                                 [ 65%]
test_manager_endpoints.tavern.yaml::GET /manager/stats/weekly PASSED                                                 [ 68%]
test_manager_endpoints.tavern.yaml::GET /manager/stats/analysisd PASSED                                              [ 71%]
test_manager_endpoints.tavern.yaml::GET /manager/stats/remoted PASSED                                                [ 75%]
test_manager_endpoints.tavern.yaml::GET /manager/logs PASSED                                                         [ 78%]
test_manager_endpoints.tavern.yaml::GET /manager/logs/summary PASSED                                                 [ 81%]
test_manager_endpoints.tavern.yaml::GET /manager/api/config PASSED                                                   [ 84%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration/validation (OK) PASSED                                [ 87%]
test_manager_endpoints.tavern.yaml::GET /manager/validation (KO) PASSED                                              [ 90%]
test_manager_endpoints.tavern.yaml::GET /manager/configuration/{component}/{configuration} PASSED                    [ 93%]
test_manager_endpoints.tavern.yaml::PUT /manager/configuration PASSED                                                [ 96%]
test_manager_endpoints.tavern.yaml::PUT /manager/restart PASSED                                                      [100%]

======================================== 32 passed, 2 warnings in 89.26s (0:01:29) =========================================
```
</details>

### CODE REVIEW

**Comment by Desvelao** [comment]
Hi, we need to add the `distinct` query parameter to

```
GET /syscollector/{agent_id}/netaddr
```

I added the requirement to the issue: https://github.com/wazuh/wazuh/issues/16715#issuecomment-1534332052

**Comment by Desvelao** [comment]
Please, review this message: https://github.com/wazuh/wazuh/issues/16715#issuecomment-1534410555

File: framework/wazuh/core/utils.py
Comment (GGP1) [inline framework/wazuh/core/utils.py:157]: Performance-wise it's better to use empty lists `[]` instead of `list()` (https://stackoverflow.com/a/5790954)

**Review by RamosFe — APPROVED** [review]

**Comment by RamosFe** [comment]
## LGTM 🟢 
The unit tests passed without any problems and the implemented changes comply with the requirements Well done!

**Comment by mhamra** [comment]
### REVISION
- branch comparison. The branch has some commits behind the master branch; one introduces changes to the spec.yaml file.
- functional requirements validation against code changes in spec.yaml

Endpoint | Add distinct | spec.yaml line | comment
-- | -- | -- | --
GET   /agents | yes | 7215 |  
GET /lists | yes | 11346 |  
GET /decoders | yes | 14662 |  
GET /decoders/files | yes | 14736 |  
GET   /manager/configuration?section=ruleset&field=decoder_dir | yes |   | not required see issue 16715
GET /groups | yes | 8318 |  
GET /groups/{group_id}/agents | yes | 8429 |  
GET /groups/{group_id}/files | yes | 8673 |  
GET /rules | yes | 13766 |  
GET /rules/files | yes | 13954 |  
GET   /manager/configuration?section=ruleset&field=rule_dir | yes |   | not required see issue 16715
GET /mitre/groups | yes | 13047 |  
GET /mitre/mitigations | yes | 13192 |  
GET /mitre/software | yes | 13311 |  
GET /mitre/tactics | yes | 13383 |  
GET /mitre/techniques | yes | 13468 |  
GET /sca/{agent_id} | yes | 14271 | already in master, not changed by this version
GET /syscollector/{agent_id}/netiface | yes | - | issue 16715 includes this endpoint, but the parameter was not found in   yaml
GET /syscollector/{agent_id}/ports | yes | 16659 |  
GET /syscollector/{agent_id}/netaddr | yes | 16284 | Change made to endpoint but not marked in issue 16715
GET /syscollector/{agent_id}/netproto | yes | 16429 |  
GET /syscollector/{agent_id}/packages | yes | 16555 |  
GET /syscollector/{agent_id}/processes | yes | 16770 |  
GET /syscollector/{agent_id}/hotfixes | yes | 16223 |  
GET /mitre/tactics | yes | 13383 |  
GET /mitre/techniques | yes | 13468 |  
GET /security/policies | yes | 18098 |  
GET /security/roles | yes | 17583 |  
GET /security/role_mapping | yes | not found endpoint | issue 16715 refers this endpoint as not found
GET /security/users | yes | 17317 |  
GET /manager/logs | yes | 12711 |  
GET /cluster/{node_id}/logs | yes | 11073 |  
GET /cluster/nodes | yes | 9533 |

**Comment by GGP1** [comment]
@mhamra just to confirm:

- There is no request to add the `distinct` parameter to the endpoints `GET /syscollector/{agent_id}/netiface` and `GET /syscollector/{agent_id}/netaddr`. We could add it if we believe all `GET /syscollector/{agent_id}/...` endpoints should have the same parameters.
- The endpoint `GET /security/role_mapping` does not exist and it was skipped.

File: framework/wazuh/manager.py
Comment (mhamra) [inline framework/wazuh/manager.py:287]: I think this docstring is wrong. There is no distinct parameter in this function.

Comment (GGP1) [inline framework/wazuh/manager.py:287]:
```suggestion
```

**Review by mhamra — DISMISSED** [review]
After resolving the observations documented in this [comment](https://github.com/wazuh/wazuh/pull/16867#issuecomment-1570560040), I approve this PR.

### **LGTM!**

Thanks @GGP1

**Comment by GGP1** [comment]
## Checks

The `Integration API endpoints test` check passed in the previous run and the only changes I introduced in this one were on docstrings (no funcional changes), thus, the check should pass again now. I'm re-running it so it passes.

**Comment by Desvelao** [comment]
> @mhamra just to confirm:
> 
> * `/syscollector/{agent_id}/netaddr` was requested after the initial development and I forgot to check the box.
> * The endpoint `GET /security/role_mapping` does not exist and it was skipped.

I think I missed the endpoint request. I reviewed it and this is `GET /security/rules` instead.

**Review by Desvelao — CHANGES_REQUESTED** [review]
I think I missed the endpoint request. I reviewed it and this is GET /security/rules instead.

Please, if you find some error in the request, notify me before resolving the pull request or closing the issue as done.

**Review by RamosFe — DISMISSED** [review]
The changes proposed by @mhamra  observations looks good.
LGTM !

File: framework/wazuh/core/tests/test_utils.py
Comment (fdalmaup) [inline framework/wazuh/core/tests/test_utils.py:220]: Since `distinct` is of type `bool`, consider defining `False` instead of `None` for false cases, or are there any cases where `process_array` is called with `None`?

**Review by fdalmaup — CHANGES_REQUESTED** [review]
Due to the introduced changes, please consider adding the output of the `framework` unit tests and the modified AITs in the issue or in the PR description, especially the latter ones since the checks are not consistent although they all passed when launched locally.

File: framework/wazuh/core/tests/test_utils.py
Comment (GGP1) [reply framework/wazuh/core/tests/test_utils.py:220]: > are there any cases where process_array is called with None?

No, all `distinct` default values were added as `False` and I changed the existing ones from `None` to `False` as well. I will change the ones from the test as suggested.

**Review by Selutario — APPROVED** [review]
LGTM, moving it to blocked until merging https://github.com/wazuh/wazuh/pull/17703 and rebasing.

**Review by mhamra — COMMENTED** [review]
Line 163 seeks the element, comparing the searched item with all list elements. In large lists, this could be inefficient. Consider using a dictionary to search distinct elements.

https://github.com/wazuh/wazuh/blob/f7f4335a9e3ee823b4469dd308f46a4c01dfddd1/framework/wazuh/core/utils.py#L156-L174

**Comment by Selutario** [comment]
Done in https://github.com/wazuh/wazuh/pull/17865

**Pull request closed without merging**
